common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RetryingBlockTransferorSuite.java

```diff
@@ -341,6 +343,35 @@ public void testBlockTransferFailureAfterSasl() throws IOException, InterruptedE
     assert(_retryingBlockTransferor.getRetryCount() == 1);
   }
 
+  @Test
+  public void testIOExceptionFailsConnectionEvenWithSaslException()
```

Comment:
do you think we should also expose `saslRetryCount` as `@VisibleForTesting` and assert the count in the tests?

Reply:
See the change w.r.t. your 3rd comment.
With that change, I think we don't need to expose `saslRetryCount`.

In the future, we can use a map (type of exception -> outstanding retry count for that type) where we can think more about asserting on retry count for each type.

Reply:
makes sense. I had a similar idea here: https://github.com/apache/spark/pull/38959/commits/3440f1259e7a04ac0f38ece8b1ce69a9253d43b7
But we ended up opting to go with a simple flag for now. I agree with using a map in the future where we might have retry counts for each type.